core/sql/optimizer/NATable.cpp

```diff
@@ -3702,25 +3706,47 @@ NAType* getSQColTypeForHive(const char* hiveType, NAMemory* heap)
       else
        continue;
     }
+
     if(pend - pstart > 0)
     {
+      if( (pend - pstart) >= MAX_NUM_LEN ) // too long
+        return NULL;
       strncpy(pstr,hiveType+pstart, pend-pstart);
       p=atoi(pstr);
     }
+
     if(send - sstart > 0)
     {
+      if( (send - sstart) >= MAX_NUM_LEN ) // too long
+        return NULL;
       strncpy(sstr,hiveType+sstart,send-sstart);
       s=atoi(sstr);
     }
 
-    if( (p>0) && (s>0) )
+    if( (p>0) && (p <= MAX_PRECISION_ALLOWED) ) //have precision between 1 - 18
     {
-      return new (heap) SQLDecimal( p, s, TRUE, TRUE);
+      if( ( s >=0 )  &&  ( s<= MAX_SCALE_ALLOWED) ) //have valid scale
+        return new (heap) SQLDecimal( p, s, TRUE, TRUE);
+      else
+        return NULL;
     }
-    else
+    else if( (p > MAX_PRECISION_ALLOWED)  && ( p <= HIVE_MAX_PRECISION_ALLOWED) )  
     {
+      if ( (s>=0) && ( s< p ) ) //have valid scale
```

Comment:
I think this could check for s <= p. Making scale and precision the same should be allowed.